### PR TITLE
Adding new filter rule(Ubuntu 20.04 does not support CUDA <10.2)

### PR DESCRIPTION
**Pull request by YuriiPerets**
fix: #32

### CODE REVIEW

**Comment by SimeonEhrig** [comment]
I did a mistake in the issue description. CUDA 10.2 and older is not available in Ubuntu 20.04. You can check for `< pkv.parse("11.0)`.

File: src/bashi/filter_software_dependency.py [hunk 1]
```diff
@@ -104,4 +115,36 @@ def software_dependency_filter(
...
+                    "is not available in Ubuntu "
+                    f"{__ubuntu_version_to_string(row[UBUNTU].version)}",
+                )
+                return False
+        if DEVICE_COMPILER in row and row[DEVICE_COMPILER].name == NVCC:
+            if row[DEVICE_COMPILER].version < pkv.parse("10.2"):
+                return False
+        for compiler in (HOST_COMPILER, DEVICE_COMPILER):
+            if compiler in row and row[compiler].name == CLANG_CUDA:
+                if row[compiler].version < CLANG_CUDA_MAX_CUDA_VERSION[3].clang_cuda:
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:147]: Good point, that you check also for `Clang-CUDA` :-) Thanks correct and didn't thought about it.

Unfortunately, there is no guaranty that `CLANG_CUDA_MAX_CUDA_VERSION` keeps his ordering. Therefore you need to replace the direct access with a search function. Otherwise your code could silently break, if somebody modifies the `CLANG_CUDA_MAX_CUDA_VERSION` list. You need to implement the following function:

```python
def get_max_supported_clang_cuda_ver(cuda_ver : str) -> str:
  """ returns the latest supported clang-cuda version for a given cuda version"
  # implement search algorithm 
  # don't expect specific ordering of the list CLANG_CUDA_MAX_CUDA_VERSION 
  # implement also tests 
```
Implement the function in the `version.py`.

By the way, an entry of `CLANG_CUDA_MAX_CUDA_VERSION` stores only string. Therefore you need to parse the version. At the moment, you compare version objects with strings, which is always wrong.
https://github.com/alpaka-group/bashi/blob/cc789bd988db738d39c5a71b16f3523ce5906734/src/bashi/versions.py#L51-L64

File: src/bashi/results.py [hunk 2]
```diff
@@ -792,3 +793,31 @@ def _remove_unsupported_cmake_versions_for_clangcuda(
             value_name2=CMAKE,
             value_version2=">3.18",
         )
+
+
+def _remove_unsupported_cuda_versions_for_ubuntu(
```
Comment (SimeonEhrig) [inline src/bashi/results.py:798]: This function needs also to remove unsupported `device compiler nvcc + ubuntu 18.04` combinations and `clang-cuda + ubuntu 18.04`. I saw that the example.py misses the following combinations:

```bash
ParameterValuePair(first=ParameterValueSingle(parameter='device_compiler', parameterValue=ParameterValue(name='nvcc', version=<Version('10.0')>)), second=ParameterValueSingle(parameter='ubuntu', parameterValue=ParameterValue(name='ubuntu', version=<Version('20.4')>))) is missing in combination list
ParameterValuePair(first=ParameterValueSingle(parameter='device_compiler', parameterValue=ParameterValue(name='nvcc', version=<Version('10.1')>)), second=ParameterValueSingle(parameter='ubuntu', parameterValue=ParameterValue(name='ubuntu', version=<Version('20.4')>))) is missing in combination list
```

But we cannot install CUDA 10.0 and 10.1 on Ubuntu 20.04

**Comment by SimeonEhrig** [comment]
I strongly recommend to fix the function `_remove_unsupported_cuda_versions_for_ubuntu()` first. You can do it with test cases and it is much easier than writing the filter rules. One time, if the function is working, running the `example.py` gives you also the correct hints, what is not correctly filtered and which filter rules needs to be fixed.

For example the `jobs.yml` which is generated by the `example.py` still the combination Ubuntu 18.04 + CUDA 10.1 if the  host compiler is Clang-CUDA. We do not expect the combination. But `_remove_unsupported_cuda_versions_for_ubuntu()` is not correctly working and therefore the `example.py` will not tell you, that this unexpected combination was generated.

File: src/bashi/filter_software_dependency.py (hunk 1 above)
Comment (SimeonEhrig) [reply src/bashi/filter_software_dependency.py:154]: I suggest that the interface takes a string and the first step is parsing it to a version object. Makes the function call easier:

```python
# expected string as argument (preferred version)
get_max_supported_clang_cuda_ver("11.2")

# expected version object as argument
get_max_supported_clang_cuda_ver(pkv.parse("11.2")
```

```python
def get_max_supported_clang_cuda_ver(cuda_ver : str) -> str:
  """ returns the latest supported clang-cuda version for a given cuda version"
  parsed_cuda_ver: pkv.Version = pkv.parse(cuda_ver)
  # implement search algorithm 
  # don't expect specific ordering of the list CLANG_CUDA_MAX_CUDA_VERSION 
  # implement also tests 
  ```

File: src/bashi/results.py [hunk 3]
```diff
@@ -827,3 +833,47 @@ def _remove_all_rocm_images_older_than_ubuntu2004_based(
...
+    parameter_value_pairs_ranges: List[ParameterValuePair],
+    removed_parameter_value_pairs: List[ParameterValuePair],
+):
+    removed_parameter_value_pairs(
+        parameter_value_pairs_ranges,
+        removed_parameter_value_pairs,
```
Comment (SimeonEhrig) [inline src/bashi/results.py:844]: Instead calling the new function, you simply renamed the variable.

```suggestion
    parameter_value_pairs: List[ParameterValuePair],
    removed_parameter_value_pairs: List[ParameterValuePair],
):
    remove_parameter_value_pairs_ranges(
        parameter_value_pairs,
        removed_parameter_value_pairs,
```

File: src/bashi/results.py [hunk 4]
```diff
@@ -827,3 +833,56 @@ def _remove_all_rocm_images_older_than_ubuntu2004_based(
...
+    removed_parameter_value_pairs: List[ParameterValuePair],
+):
+    remove_parameter_value_pairs_ranges(
+        parameter_value_pairs,
+        removed_parameter_value_pairs,
+        parameter1=UBUNTU,
+        value_name1=UBUNTU,
+        value_min_version1=20.04,
+        value_min_version1_inclusive=True,
+        value_max_version1=9999,
```
Comment (SimeonEhrig) [inline src/bashi/results.py:849]: You can remove this argument. If you use the default argument for `value_max_version1`, it will delete all version bigger than 20.04, including 20,04.

File: src/bashi/results.py [hunk 5]
```diff
@@ -827,3 +833,56 @@ def _remove_all_rocm_images_older_than_ubuntu2004_based(
...
+    )
+
+    remove_parameter_value_pairs_ranges(
+        parameter_value_pairs,
+        removed_parameter_value_pairs,
+        parameter1=UBUNTU,
+        value_name1=UBUNTU,
+        value_min_version1=20.04,
+        value_min_version1_inclusive=True,
+        value_max_version1=9999,
```
Comment (SimeonEhrig) [inline src/bashi/results.py:865]: Use the default argument

File: src/bashi/results.py [hunk 6]
```diff
@@ -827,3 +833,56 @@ def _remove_all_rocm_images_older_than_ubuntu2004_based(
...
+    )
+    for compiler_type in (HOST_COMPILER, DEVICE_COMPILER):
+        remove_parameter_value_pairs_ranges(
+            parameter_value_pairs,
+            removed_parameter_value_pairs,
+            parameter1=UBUNTU,
+            value_name1=UBUNTU,
+            value_min_version1=20.04,
+            value_min_version1_inclusive=True,
+            value_max_version1=9999,
```
Comment (SimeonEhrig) [inline src/bashi/results.py:881]: Use the default argument

File: tests/test_results.py [hunk 7]
```diff
@@ -2492,3 +2493,226 @@ def test_remove_all_rocm_images_older_than_ubuntu2004_based(self):
...
+                ),
+                OD(
+                    {
+                        UBUNTU: (UBUNTU, "20.04"),
+                        DEVICE_COMPILER: (NVCC, "10.2"),
+                    }
+                ),
+            ]
+        )
+        expected_results = parse_expected_val_pairs(
```
Comment (SimeonEhrig) [inline tests/test_results.py:2630]: Was some work to review this, but the expected results looks correct and the test in general. Remove the the `value_max_version1=9999,` and the validation site is ready. Therefore only the filter rules needs to be fixed.

File: src/bashi/results.py [hunk 8]
```diff
@@ -827,3 +833,56 @@ def _remove_all_rocm_images_older_than_ubuntu2004_based(
...
+        value_name1=UBUNTU,
+        value_min_version1=20.04,
+        value_min_version1_inclusive=True,
+        value_max_version1=9999,
+        parameter2=ALPAKA_ACC_GPU_CUDA_ENABLE,
+        value_name2=ALPAKA_ACC_GPU_CUDA_ENABLE,
+        value_min_version2=OFF,
+        value_min_version2_inclusive=False,
+        value_max_version2=11,
+        value_max_version2_inclusive=True,
```
Comment (SimeonEhrig) [inline src/bashi/results.py:855]:
```suggestion
        value_max_version2_inclusive=False,
```
Needs to be exclusive, because otherwise you remove the pair Ubuntu 20.04 + CUDA 11.0.
Please also the pair to the tests. This corner case is not covered by your test casese. Only by the long running test.

File: tests/test_filter_software_dependency.py [hunk 9]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                    ),
+                )
+            )
+
+        for nvcc_version in [2, 6, 10.1, 10.2, 11.4, 15]:
+            self.assertTrue(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, nvcc_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:353]:
```suggestion
                            DEVICE_COMPILER: ppv((NVCC, nvcc_version)),
```

File: tests/test_filter_software_dependency.py [hunk 10]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                        }
+                    ),
+                )
+            )
+        for nvcc_version in [11.1, 11.2, 11.4, 15]:
+            self.assertTrue(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, nvcc_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:364]:
```suggestion
                            DEVICE_COMPILER: ppv((NVCC, nvcc_version)),
```

File: tests/test_filter_software_dependency.py [hunk 11]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:375]:
```suggestion
                            DEVICE_COMPILER: ppv((NVCC, nvcc_version)),
```

File: tests/test_filter_software_dependency.py [hunk 12]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                        }
+                    ),
+                )
+            )
+        for clang_cuda_version in [2, 6, 10.1, 10.2, 11, 12.4, 15]:
+            self.assertTrue(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:386]:
```suggestion
                            DEVICE_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 13]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                        }
+                    ),
+                )
+            )
+        for clang_cuda_version in [2, 6, 10.1, 10.2, 11]:
+            self.assertTrue(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:397]:
```suggestion
                            DEVICE_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 14]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:408]:
```suggestion
                            DEVICE_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 15]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                        }
+                    ),
+                )
+            )
+        for clang_cuda_version in [2, 6, 10.1, 10.2, 11, 12.4, 15]:
+            self.assertTrue(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            HOST_COMPILER: ppv((HOST_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:419]:
```suggestion
                            HOST_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 16]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                        }
+                    ),
+                )
+            )
+        for clang_cuda_version in [2, 6, 10.1, 10.2, 11]:
+            self.assertTrue(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            HOST_COMPILER: ppv((HOST_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:430]:
```suggestion
                            HOST_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 17]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:441]:
```suggestion
                            HOST_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 18]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                reason_msg.getvalue(),
+                f"CUDA {CUDA_version} is not available in Ubuntu 22.04",
+            )
+        for nvcc_version in [1, 6, 10.1]:
+            reason_msg = io.StringIO()
+            self.assertFalse(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, nvcc_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:495]:
```suggestion
                            DEVICE_COMPILER: ppv((NVCC, nvcc_version)),
```

File: tests/test_filter_software_dependency.py [hunk 19]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                reason_msg.getvalue(),
+                f"NVCC {nvcc_version} is not available in Ubuntu 20.04",
+            )
+        for nvcc_version in [1, 6, 10.1]:
+            reason_msg = io.StringIO()
+            self.assertFalse(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, nvcc_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:513]:
```suggestion
                            DEVICE_COMPILER: ppv((NVCC, nvcc_version)),
```

File: tests/test_filter_software_dependency.py [hunk 20]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                f"NVCC {nvcc_version} is not available in Ubuntu 22.04",
+            )
+
+        for clang_cuda_version in [1, 6, 10.1]:
+            reason_msg = io.StringIO()
+            self.assertFalse(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:532]:
```suggestion
                            DEVICE_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 21]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                reason_msg.getvalue(),
+                f"clang cuda {clang_cuda_version} is not available in Ubuntu 22.04",
+            )
+        for clang_cuda_version in [1, 6, 10.1]:
+            reason_msg = io.StringIO()
+            self.assertFalse(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            HOST_COMPILER: ppv((HOST_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:550]:
```suggestion
                            HOST_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 22]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                f"clang cuda {clang_cuda_version} is not available in Ubuntu 22.04",
+            )
+
+        for clang_cuda_version in [1, 6, 10.1]:
+            reason_msg = io.StringIO()
+            self.assertFalse(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            DEVICE_COMPILER: ppv((DEVICE_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:569]:
```suggestion
                            DEVICE_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

File: tests/test_filter_software_dependency.py [hunk 23]
```diff
@@ -302,3 +302,297 @@ def test_non_valid_ROCm_images_Ubuntu2004_based_d4(self):
...
+                reason_msg.getvalue(),
+                f"clang cuda {clang_cuda_version} is not available in Ubuntu 20.04",
+            )
+        for clang_cuda_version in [1, 6, 10.1]:
+            reason_msg = io.StringIO()
+            self.assertFalse(
+                software_dependency_filter_typechecked(
+                    OD(
+                        {
+                            HOST_COMPILER: ppv((HOST_COMPILER, clang_cuda_version)),
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:587]:
```suggestion
                            HOST_COMPILER: ppv((CLANG_CUDA, clang_cuda_version)),
```

**Comment by SimeonEhrig** [comment]
@YuriiPerets  You did the same mistake with the host or device compiler parameter-value-pair many times. Please check also by yourself, if I missed to marked it in your code somewhere and fix it by yourself.

File: src/bashi/filter_software_dependency.py [hunk 24]
```diff
@@ -116,4 +117,39 @@ def software_dependency_filter(
...
+                )
+                return False
+        for compiler_type in (HOST_COMPILER, DEVICE_COMPILER):
+            if compiler_type in row and row[compiler_type].name == CLANG_CUDA:
+                if row[compiler_type].version < get_oldest_supporting_clang_version_for_cuda(
+                    "11.0"
+                ):
+                    reason(
+                        output,
+                        f"{__pretty_name_compiler(compiler_type)} clang cuda {row[compiler_type].version} "
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:150]:
```suggestion
                        f"{__pretty_name_compiler(compiler_type)} clang-cuda {row[compiler_type].version} "
```
Just a little bit sugar ;-)

**Comment by SimeonEhrig** [comment]
@YuriiPerets Nice work. Please fix this one reason message, than squash and rebase please. Hopefully if you did everything and force push, the CI will be triggered again. I executed the test locally and it looks good.

**Review by SimeonEhrig — APPROVED** [review]

**Review by SimeonEhrig — APPROVED** [review]
